compilers/concrete-optimizer/concrete-optimizer/src/optimization/dag/multi_parameters/optimize/mod.rs

```diff
@@ -79,6 +141,8 @@ type FksSrc = PartitionIndex;
 
 #[inline(never)]
 fn optimize_1_ks(
+    search_space_restriction: &impl SearchSpaceRestriction,
```

Comment:
could be optional so it's optimized away. not sure how this is compiled with impl. Is it inlined for each type ?

Reply:
guess is noop

Reply:
So, `impl SearchSpaceRestriction` is just equivalent to `<T: SearchSpaceRestriction>`. It is a generic parameter in disguise. This means that the function gets monomorphized for every `T` it is called with. When it is called with `NoRestriction`, every call to `is_available` returns the constant true, so yeah, likely the check is completely removed. 

As a side note, this is much better than using an `Option`, which would check if the option contains something at every call to `is_available`.